### DISCUSSION TITLE
hook malloc  测试首次正常，后面的不会回调（小米13）

### DISCUSSION THREAD

**Discussion by HarlonWang** [bytedance/android-inline-hook#76] [Q&A] [answered]
以下代码，在小米13上测试下来，第二行及以后的 malloc 的调用，hook 那里都不会被触发：
```c
typedef void *(*type_malloc)(size_t);
void *orig_malloc;
void* my_malloc(size_t size) {
    void* p = ((type_malloc)orig_malloc)(size);
    __android_log_print(ANDROID_LOG_ERROR, "shadowhook_tag", "内存分配: %zu 字节, 地址: %p", size, p);
    return p;
}

extern "C" JNIEXPORT jstring JNICALL
Java_com_example_myapplication_MainActivity_stringFromJNI(
        JNIEnv* env,
        jobject /* this */) {
    shadowhook_init(SHADOWHOOK_MODE_SHARED, true);

    void *stub = shadowhook_hook_sym_addr((void *)malloc, (void *)my_malloc, &orig_malloc);
    if(stub == NULL)
    {
        int error_num = shadowhook_get_errno();
        const char *error_msg = shadowhook_to_errmsg(error_num);
        __android_log_print(ANDROID_LOG_WARN,  "test", "hook failed: %d - %s", error_num, error_msg);
    }
    // 这行会触发
    malloc(123);
   // 这以后都不会触发
    malloc(321);
    malloc(432);
    malloc(543);
    ...
```
日志如下：
![image](https://github.com/user-attachments/assets/907dd2a9-d15b-41b2-b877-bcf1719dea8b)

帮忙看下，是不是我的代理函数，使用姿势不对 :(

**Comment by Colibrow** [comment, marked as answer]
https://github.com/bytedance/android-inline-hook/blob/main/doc/manual.zh-CN.md#%E4%B8%BE%E4%BE%8B%E4%B8%80c%E6%BA%90%E6%96%87%E4%BB%B6
在SHARED模式下代理函数中需要执行SHADOWHOOK_POP_STACK

**Reply by HarlonWang** [reply]
> https://github.com/bytedance/android-inline-hook/blob/main/doc/manual.zh-CN.md#%E4%B8%BE%E4%BE%8B%E4%B8%80c%E6%BA%90%E6%96%87%E4%BB%B6 在SHARED模式下代理函数中需要执行SHADOWHOOK_POP_STACK

感谢大佬，我尝试这样做了，但是编译报错了：
![image](https://github.com/user-attachments/assets/609c02ea-bafd-4a48-9ea0-b022684cc5b9)

你有遇到过吗？

**Reply by HarlonWang** [reply]
是只能在 c 文件里这么用么，cpp 文件不能直接这样用吗

**Reply by HarlonWang** [reply]
忽略吧，看起来是编译环境的问题，我改成 c 文件，可以编译通过了

**Reply by Colibrow** [reply]
cpp需要根据示例二来写代理函数

**Reply by HarlonWang** [reply]
好的🙏，看到说明了
![image](https://github.com/user-attachments/assets/63f7bbf1-4fcb-47dc-bb78-6bc257b8479d)
